### PR TITLE
[Coupons] Edit expiry date

### LINKED ISSUE #6494: [Coupons] Add option to change the expiration date
Add option to show a date picker to set an expiration date:
Screen Shot 2022-05-11 at 12 01 49

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #6494 

### Description
This PR adds support for editing the expiry date.
For the date picker, I had two options:
1. Use the MaterialDatePicker from the Fragment using events.
2. Create a Compsable date picker.

I went with the second approach, because it was a nice challenge, and also because Google is working on composable [implementations](https://issuetracker.google.com/issues/197942880), which means migrating to them would be easier.

Another point, regarding the scenario for deleting the expiry date, I went with a Dialog that shows two buttons for selecting the action (edit vs delete), this aligns with the iOS implementation (they use an action sheet).

### Testing instructions
1. Make sure coupons beta switch is enabled.
2. Open details of a coupon in the app, then navigate to edit using the menu.
3. Click on the expiry date.
4. If the date wasn't set, confirm that a picker is displayed.
5. If the date was already set, confirm that a dialog is displayed with two buttons.
6. Change the date, and confirm the changes are reflected.
7. Delete the date, and confirm the picker changes to "None"

### Images/gif
https://user-images.githubusercontent.com/1657201/168159961-6e15fcc6-4303-49f8-8ee6-edfae5f93e02.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by hafizrahman — CHANGES_REQUESTED** [review]
A few initial thoughts and discoveries while testing:

I think it's cool you can recreate the date picker in Compose pretty quickly! Shows how powerful Compose is. 👍🏼  ✨ 

Currently, if the calendar dialog is opened and the orientation is changed, the calendar disappears.

Next, when using "Large" system font setting and more, coupled with the landscape orientation, the buttons at the bottom is cut off:
Screenshot_20220516_132707

Next, the two-buttons "Edit Expiry" and "Delete Expiry" dialog felt a little unusual to me, mostly because I have not seen a lot of similar dialog in Android apps:

Screenshot_20220516_133923


I then tried to figure out what the equivalent of iOS Action Sheet for Android. [This SO conversation](https://stackoverflow.com/questions/47645736/android-equivalence-of-ios-actionsheet) recommends either using Bottom Sheet Dialog, or an [Alert Dialog with list](https://developer.android.com/guide/topics/ui/dialogs#AddingAList). In the old [Material Design for iOS documentation for Action Sheet](https://material.io/develop/ios/supporting/action-sheet), the design also looks a lot like Bottom Sheet Dialog. Or maybe we can use a dialog with three options (Cancel, Edit, Delete)?

Overall I think it's not a big deal, it still works and probably worth re-iterating later. Curious what @woocommerce/mobile-designers think about it.

**Comment by hichamboushaba** [comment]
Thanks @hafizrahman for the review 👍 

> Currently, if the calendar dialog is opened and the orientation is changed, the calendar disappears.

This is something that happens in many places in the app, but since we are using Compose now, and saving things to the SavedState is easier, I agree with you that we should implement it, it's fixed now with the two commits: [af5754a](https://github.com/woocommerce/woocommerce-android/pull/6505/commits/af5754ab2b3643ee6572cc8b686870d2cc8c5f6c) and [5187091](https://github.com/woocommerce/woocommerce-android/pull/6505/commits/51870914eb962d7ddc012e189ec6513b624b46dc)

> Next, when using "Large" system font setting and more, coupled with the landscape orientation, the buttons at the bottom is cut off

Nice catch, I spent some time optimizing the layout to be more responsive, and it was a good exercise in learning how [IntrinsicSize](https://developer.android.com/jetpack/compose/layouts/intrinsic-measurements) works in Compose.
I tried to have a design that matches what the `MaterialDatePicker` does:
| portrait | landscape |
| ---- | ---- |
| Screen Shot 2022-05-16 at 15 53 30 | Screen Shot 2022-05-16 at 15 53 14 |
| Screen Shot 2022-05-16 at 15 32 48 | Screen Shot 2022-05-16 at 15 33 23 |
| Screen Shot 2022-05-16 at 15 29 28 | Screen Shot 2022-05-16 at 15 29 49 |

> I then tried to figure out what the equivalent of iOS Action Sheet for Android

It's not about direct equivalent. Yes, we have a bottom sheet in Android, but we don't use them for this kind of interaction in the app, if you check for example the `Image type picker` in the app, you'll see that we use a dialog in Android, while iOS uses an action sheet, as it's the natural thing for both platforms.
This being said, the design is temporary, and I'm already waiting for the designer's feedback p1652435227178049-slack-C02TD63FZA8, and as this is behind a feature flag, I believe it's OK to merge it, and update the design later when we get a final response.

File: WooCommerce/build.gradle [hunk 1]
```diff
@@ -107,6 +107,9 @@ android {
 
     kotlinOptions {
         jvmTarget = "1.8"
+        freeCompilerArgs += [
+                "-Xopt-in=kotlin.RequiresOptIn"
+        ]
```
Comment (hichamboushaba) [inline WooCommerce/build.gradle:112]: This allows using the `OptIn` annotation which is better than declaring components as experimental, and ending with tons of warnings.

For this case, the clients don't have to care that the `DatePickerDialog` is using an experimental API, so when the component opts in, callers will be able to call it without any special setup.

**Review by hafizrahman — CHANGES_REQUESTED** [review]
I'm very impressed by the landspace mode redesign. Awesome and thanks for implementing that! It's great that it's now preserved on config change as well.

A couple of things that I just noticed:

1. If I go to a previous month and select a date, the month has this swipe animation but then opens to the same month. That animation does not feel necessary to me (since it opens to the same month). Video here to show what I meant:


https://user-images.githubusercontent.com/266376/168717393-f8b60334-dad2-4a2c-9047-f88ed780b1f5.mov

[In Material date picker documentation](https://material.io/components/date-pickers), it says that:

> Accessibility
> On mobile, a date picker should allow users to enter dates manually by inputting text, without using the picker. This interaction allows multiple input methods and makes it accessible to those using assistive technology.
> On mobile, the input picker should be available from the calendar pickers via the edit icon.

Do you think this is something we can implement right away? I can also make a separate issue if we want to go forward without it first.

**Comment by hichamboushaba** [comment]
Thanks @hafizrahman 
> If I go to a previous month and select a date

Nice catch, I'll try to investigate why it happens, but IMO it's not a blocker 🙂.

> Do you think this is something we can implement right away?

I don't think we should implement it for the coupons project, the component follows the same behavior as `MaterialDatePicker`, which doesn't offer this possibility.
This is something we should think about for the whole app, and if we find it necessary, then it should be done for all parts where we use date pickers (like Products' sale scheduling).

**Review by hafizrahman — APPROVED** [review]
Thank you for addressing my points! This is good to go now I feel (aside from the existing conflicts) 👍🏼 👍🏼

**Comment by hichamboushaba** [comment]
Thanks @hafizrahman for the review, I'll merge this as, and I'll have a separate PR for the dialog adjustments to align with the design changes.

**Pull request merged**